File: java/aircraft-positions/src/main/java/com/thehecklers/aircraftpositions/WebSocketHandler.java
package com.thehecklers.aircraftpositions;

import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.CloseStatus;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;
import org.springframework.web.socket.handler.TextWebSocketHandler;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@RequiredArgsConstructor
@Component
public class WebSocketHandler extends TextWebSocketHandler {
    private final List<WebSocketSession> sessionList = new ArrayList<>();
    @NonNull
    private final AircraftRepository repository;

    public List<WebSocketSession> getSessionList() {
        return sessionList;
    }

    @Override
    public void afterConnectionEstablished(WebSocketSession session) throws Exception {
        sessionList.add(session);
        System.out.println("Connection established from " + session.toString() +
                " @ " + Instant.now().toString());
    }

    @Override
    protected void handleTextMessage(WebSocketSession session, TextMessage message)
            throws Exception {
        try {
            System.out.println("Message received: '" + message + "', from " +
                    session.toString());

            for (WebSocketSession sessionInList : sessionList) {
                if (sessionInList != session) {
                    sessionInList.sendMessage(message);
                    System.out.println("--> Sending message '" + message + "' to " +
                            sessionInList.toString());
                }
            }
        } catch (Exception e) {
            System.out.println("Exception handling message: " + e.getLocalizedMessage());
        }
    }

    @Override
    public void afterConnectionClosed(WebSocketSession session, CloseStatus status)
            throws Exception {
        sessionList.remove(session);
        System.out.println("Connection closed by " + session.toString() +
                " @ " + Instant.now().toString());
    }
}

File: kotlin/aircraft-positions/src/main/kotlin/com/thehecklers/aircraftpositions/WebSocketHandler.kt
package com.thehecklers.aircraftpositions

import org.springframework.stereotype.Component
import org.springframework.web.socket.CloseStatus
import org.springframework.web.socket.TextMessage
import org.springframework.web.socket.WebSocketSession
import org.springframework.web.socket.handler.TextWebSocketHandler
import java.time.Instant
import java.util.*


@Component
class WebSocketHandler(private val repository: AircraftRepository) : TextWebSocketHandler() {
    private val sessionList = ArrayList<WebSocketSession>()

    fun getSessionList() = sessionList

    @Throws(Exception::class)
    override fun afterConnectionEstablished(session: WebSocketSession) {
        sessionList.add(session)
        println("Connection established from $session @ ${Instant.now()}")
    }

    @Throws(Exception::class)
    override fun handleTextMessage(session: WebSocketSession, message: TextMessage) {
        try {
            println("Message received: '$message', from $session")

            for (sessionInList: WebSocketSession in sessionList) {
                if (sessionInList != session) {
                    sessionInList.sendMessage(message)
                    println("--> Sending message '$message' to $sessionInList")
                }
            }
        } catch (e: Exception) {
            println("Exception handling message: ${e.localizedMessage}")
        }
    }

    @Throws(Exception::class)
    override fun afterConnectionClosed(session: WebSocketSession, status: CloseStatus) {
        sessionList.remove(session)
        println("Connection closed by $session @ ${Instant.now()}")
    }
}